src/network/networkd-sysctl.c

```diff
@@ -5,17 +5,193 @@
 #include <linux/if_arp.h>
 
 #include "af-list.h"
+#include "cgroup-util.h"
+#include "fd-util.h"
 #include "missing_network.h"
 #include "networkd-link.h"
 #include "networkd-lldp-tx.h"
 #include "networkd-manager.h"
 #include "networkd-ndisc.h"
 #include "networkd-network.h"
 #include "networkd-sysctl.h"
+#include "path-util.h"
+#include "sd-messages.h"
 #include "socket-util.h"
 #include "string-table.h"
 #include "sysctl-util.h"
 
+#if HAVE_VMLINUX_H
+
+#include "bpf-link.h"
+
+#include "bpf/sysctl_monitor/sysctl-monitor-skel.h"
+#include "bpf/sysctl_monitor/sysctl-write-event.h"
+
+static struct sysctl_monitor_bpf *sysctl_monitor_bpf_free(struct sysctl_monitor_bpf *obj) {
+        sysctl_monitor_bpf__destroy(obj);
+        return NULL;
+}
+
+static struct ring_buffer *rb_free(struct ring_buffer *rb) {
+        sym_ring_buffer__free(rb);
+        return NULL;
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct sysctl_monitor_bpf *, sysctl_monitor_bpf_free);
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct ring_buffer *, rb_free);
+
+static int sysct_write_event(void *ctx, void *data, size_t data_sz) {
+        struct sysctl_write_event *we = ASSERT_PTR(data);
+        Hashmap **sysctl_shadow = ASSERT_PTR(ctx);
+        _cleanup_free_ char *path = NULL;
+        char *value;
+
+        if (we->version != 1)
+                return log_warning_errno(SYNTHETIC_ERRNO(EINVAL),
+                                "Invalid sysctl write event version: %d", we->version);
+
```

Comment:
and i think we should check if `we->name` is truncated or not. i.e. the bpf code should NUL terminate the string if it fits, but otherwise use the last char of the field, so that we can use this to determine truncation, and then handle this gracefully and log about the case.

Reply:
How to do it in BPF? If the name is too long, `bpf_sysctl_get_name()` truncates the value and returns `-E2BIG`.
Just set the last char to a non-null value?

Reply:
hmm so you are saying bpf_sysctl_get_name() truncates and adds a NUL byte? yuck.

i guess we could set the last byte to 0xFF as a marker for truncation?

Reply:
Yes, that's what the `bpf-helpers` manpage says.
Instead of putting markers into the string, why don't we add a `errno` member to the `sysctl_write_event` struct?
I'll set it to 0 on success, and to some error value on failure.